server.go

```diff
@@ -2604,11 +2604,12 @@ func (s *server) ConnectToPeer(addr *lnwire.NetAddress, perm bool) error {
 	// Peer was not found, continue to pursue connection with peer.
 
 	// If there's already a pending connection request for this pubkey,
-	// then we ignore this request to ensure we don't create a redundant
+	// then we cancel older requests to ensure we don't create a redundant
 	// connection.
 	if reqs, ok := s.persistentConnReqs[targetPub]; ok {
 		srvrLog.Warnf("Already have %d persistent connection "+
 			"requests for %v, connecting anyway.", len(reqs), addr)
+		s.cancelConnReqs(targetPub, nil)
```

Comment:
Log above should be updated?

Reply:
done!

Reply:
As noted above, we should only cancel them if the connection attempt was successful or if `perm` was set.

Reply:
> we should only cancel them if the connection attempt was successful

This may cause more than one connection in progress which I think will cause the remote node to drop the duplicate connection. Isn't it cleaner to make sure we only have one pending connection?



> or if `perm` was set.

Do you think there is a problem with not differentiating the "perm" case here? I think it would be nice to have the same behavior for both cases.

Thanks.

Reply:
> This may cause more than one connection in progress which I think will cause the remote node to drop the duplicate connection. Isn't it cleaner to make sure we only have one pending connection?

If we assume that the peer will drop the duplicate connection, then there shouldn't be an issue. Another alternative is to cancel the persistent request and proceed to connect. If we fail to connect, we can restore the persistent request.

> Do you think there is a problem with not differentiating the "perm" case here? I think it would be nice to have the same behavior for both cases.

Not sure what you mean by differentiating the `perm` case. If it is set we'll end up with duplicate persistent connection requests, which we want to avoid.

Reply:
> Another alternative is to cancel the persistent request and proceed to connect. If we fail to connect, we can restore the persistent request.

I think we don't need to write any code to restore the persistent request because if the new connection request fails it will try again using backoff if this peer is persistent. My intention was just to replace a pending request with another.



> Not sure what you mean by differentiating the `perm` case

Sorry I wasn't clear about it. I was saying why limiting the cancelation only for the "perm" case and what problem do you see in canceling pending requests also for non permanent connection if the user just called ConnectToPeer again?

Reply:
Let's say there was a persistent connection request to a peer because we have a channel open with them. The request is still active since the peer is offline. If the user tries to manually connect to this peer and fails, if we cancel the request without restoring it, then we wouldn't be able to reconnect to the peer once they come back online.

Reply:
Got you! , makes sense. Now I understand also why you suggested to mark persistent connections at peers.Start. This will make sure the persistent pees will survive cancellations and stay persistent.
So in summary, the problematic case is when the user tries to connect manually (perm=false) to a persistent peer with a pending connection. At this case we want to make sure the connection stays persistent regardless of the new connection attempt results.
So I think to limit the cancelation only for the cases where perm=true (as you suggested) will solve this case right?  I will update the PR.

Reply:
@wpaulino I have moved the cancelation code into the block that is executed only in case of perm connection. This intends to cover the first comment of the review (only cancel in case of perm connection).
As for the second comment (cancel pending requests if connection is successful) I saw this is already handled here: https://github.com/lightningnetwork/lnd#L2214